### ISSUE TITLE
Invalid french phone numbers

### ISSUE THREAD

**Issue by Nek-** [fzaninotto/Faker#1413]
Faker generates often invalid phone numbers for french phone numbers. The problem is that phone numbers starting by 08/07 (and maybe 09) are under serveral rules (that I don't know).

This little script always return good phone numbers:

```php
use Faker\Generator;

class Random
{
    /**
     * Faker is not able to give valid phone number every times. So here is a script
     * generating french phone numbers always valid.
     *
     * @param Generator $faker
     *
     * @return string
     */
    public static function randomPhoneNumber(Generator $faker)
    {
        $phone = '0'; // May be replaced by "+33"
        $phone .= $faker->numberBetween(1, 6); // 08 / 09 / 07 have special rules
        $phone .= $faker->regexify('[1-5]{8}');

        return $phone;
    }
}
```

**Comment by ppelgrims** [comment]
Does #1343 fix the 07 problem? So it's just 08 numbers? https://www.arcep.fr/index.php?id=8146#c9625 Would adding this, fix the problem?

How do you actually know the numbers are incorrect?

**Comment by Nek-** [comment]
I assume the other PR fix only problems for 07 phone numbers. I use this lib (which is awesome) to validate phone numbers: https://github.com/giggsey/libphonenumber-for-php

**Comment by svanpoeck** [comment]
Hi,

The fix I submitted was indeed only aimed to fixe issues with 07 mobile
numbers.

I hadn't encountered any issues generating other mobile numbers and never
have had the need to generate regular phone numbers.

We use the giggsey lib as well to validate phone numbers, I'd be interested
to know which faker-generated phone numbers fail the lib's validation.

Maybe the validation is wrong? Have you checked the arcep document Pieter
mentioned? If the faker-generated phone numbers comply with the document,
they're valid and should pass the validation.


On 31 Jan 2018 12:00 am, "Maxime Veber" <notifications@github.com> wrote:

I assume the other PR fix only problems for 07 phone numbers. I use this
lib (which is awesome) to validate phone numbers:
https://github.com/giggsey/libphonenumber-for-php

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/fzaninotto/Faker/issues/1413#issuecomment-361764061>,
or mute the thread

**Comment by ppelgrims** [comment]
I got an invalid phone number on the second try: `"+33 (0)8 50 30 52 11"`, the library correctly identifies it as [invalid](https://giggsey.com/libphonenumber/?phonenumber=08+50+30+52+11&country=FR&language=fr&region=FR) 

https://github.com/giggsey/libphonenumber-for-php/blob/master/src/data/PhoneNumberMetadata_FR.php I wonder if this is it? Lines 50, 61, and 72.

**Comment by svanpoeck** [comment]
First off, thanks for your continued efforts :-)

All lines of code pertaining to validating FR 08 numbers in giggsey are : 50, 61, 72 and 114.
The official validation source is and remains https://www.arcep.fr/index.php?id=8146#c9625, so I presume (re)writing the rules in Faker so that they comply to that would (and should) make the generated numbers validate in the giggsey library.

**Comment by ppelgrims** [comment]
They're leaving out `456` in `#61    'NationalNumberPattern' => '89[1-37-9]\\d{6}',`, I'm not really sure why? I don't think the docs say anything about that.

[giggsey](https://github.com/giggsey/libphonenumber-for-php#problems-with-invalid-numbers)'s advice is to raise it as an issue with the [libphonenumber](https://github.com/googlei18n/libphonenumber) project.

**Comment by svanpoeck** [comment]
I've no idea why they're leaving out 456 after 89, the official documentation does not mention anything special regarding the numbers that may follow "089".
I tried https://libphonenumber.appspot.com/phonenumberparser?number=08+95+23+45+56&country=FR&geocodingLocale=fr-FR and they consider it an invalid phone number as well.
I've filed an issue with them: https://issuetracker.google.com/u/1/issues/73269839

**Comment by svanpoeck** [comment]
Just a quick update: I've contacted ARCEP (the French instance for phone number regulations) and asked them if numbers 089(4|5|6) are valid. They answered that numbers 0895 are already being attributed and that 0894 numbers will be when other 089 resource ranges will come to exhaustion. Unfortunately I did not explicitly ask about 0896 numbers and therefore got no precise answer about those.
A second request to clear that out has been sent, I'm waiting for their answer.

I passed that information onto Google, and haven't heard of them back yet (which is normal since they're on a different timezone)

I'll keep the information here updated as things progress.

**Comment by ppelgrims** [comment]
Looks like it'll be fixed in the next release. Is there anything that needs to be done on our part?

**Comment by svanpoeck** [comment]
Well... according to what Google replied on https://issuetracker.google.com/issues/73269839, they'll fix (and have done so already I believe) for 0895, but not for 0894 and 0896. For the latter I'll have to sign their CLA and forward ARCEP's messages to them, which I haven't come around to yet.
I'll update here when I will have.

**Comment by svanpoeck** [comment]
Ok so I've signed their CLA and forwarded ARCEP's mail to the discussion group for the issue @Google.
I've also sent a reminder email to ARCEP because they haven't answered my question regarding 0896* numbers yet.
All of that is in Google's bug tracker at https://issuetracker.google.com/issues/73269839

**Comment by svanpoeck** [comment]
ARCEP answered me saying that 0896 is indeed to be considered as 0984: the numbers in these ranges will be attributed once the other ranges have come to congestion.

I've passed that information onto Google, and here's a copy/paste of Google's answer which I consider final:

> Thanks so much for followup on this and getting confirmation from ARCEP.
> We will add support to 894 XXX XXX and 896 XXX XXX phone numbers once we come to know that they are assigned.

So when the library https://github.com/giggsey/libphonenumber-for-php gets updated, Faker should generate phone numbers in the 0895* range, but not yet in the 0894* and 0896* ranges.

I'll send in a PR in the next few days.

**Comment by fzaninotto** [comment]
Fixed by #1439

**Comment by Nek-** [comment]
👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏
